pymuonsuite/io/castep.py

```diff
@@ -7,109 +7,348 @@
+            # have not called write() and made a calculator before.
+
+            if self._calc is None:
+                if isinstance(a.calc, Castep):
+                    self._calc = deepcopy(a.calc)
+                self._create_calculator()
+
+            self._update_calculator(calc_type)
+            a.set_calculator(self._calc)
+
+            io.write(os.path.join(folder, sname + '.cell'),
+                     a, magnetic_moments='initial')
+            write_param(os.path.join(folder, sname + '.param'),
+                        a.calc.param, force_write=True)
+
+            if self.script is not None:
+                stxt = open(self.script).read()
+                stxt = stxt.format(seedname=sname)
+                with open(os.path.join(folder, 'script.sh'), 'w') as sf:
+                    sf.write(stxt)
+        else:
+            raise(NotImplementedError("Calculation type {} is not implemented."
+                  " Please choose 'GEOM_OPT' or 'MAGRES'".format(calc_type)))
+
+    def _create_calculator(self):
+        if self._calc is not None and isinstance(self._calc, Castep):
+            calc = deepcopy(self._calc)
+        else:
+            calc = Castep()
+
+        mu_symbol = self.params.get('mu_symbol', 'H:mu')
+
+        # Start by ensuring that the muon mass and gyromagnetic ratios are
+        # included
+        gamma_block = calc.cell.species_gamma.value
+        if gamma_block is None:
+            calc.cell.species_gamma = add_to_castep_block(gamma_block,
+                                                          mu_symbol,
+                                                          constants.m_gamma,
+                                                          'gamma')
+
+            mass_block = calc.cell.species_mass.value
+            calc.cell.species_mass = add_to_castep_block(mass_block, mu_symbol,
+                                                         constants.m_mu_amu,
+                                                         'mass')
+
+        # Now assign the k-points
+        calc.cell.kpoint_mp_grid = list_to_string(
+            self.params.get('k_points_grid', [1, 1, 1]))
+
+        # Read the parameters
+        pfile = self.params.get('castep_param', None)
+        if pfile is not None:
+            calc.param = read_param(self.params['castep_param']).param
+
+        self._calc = calc
+
+        return self._calc
+
+    def _update_calculator(self, calc_type):
+        if calc_type == "MAGRES":
+            # check if our calculator is already set up for Magres
+            # if it is, we don't need to modify the calc.
+            if not self._calc.param.task == 'Magres':
+                self._create_hfine_castep_calculator()
+        elif calc_type == "GEOM_OPT":
+            # check if our calculator is already set up for geom opt
+            # if it is, we don't need to modify the calc.
+            if not self._calc.param.task == 'GeometryOptimization':
```

Comment:
So the reasoning behind this code was because when we discussed making the ReadWrite class we said that one benefit of making an instance of the class instead of having static methods was that you would save the calculator and be able to reuse it in future if you wanted to. 

So what I was trying to do was prevent us from updating a calculator when we don't need to. However, I am not sure how you could go about making sure that the calculator is unchanged so we'll probably just have to always call `_create_geom_opt_castep_calculator()`

Reply:
I have updated the code so that it always updates the calculator properties appropriately so that it is configured properly for the type of calculation if a calculator has been provided that was created outside of this class.

If the calculator that will be used for writing is already saved as a member of the class, this means it was created inside the class so has the correct properties. If the params have been modified then it will be updated with them. Otherwise, if the parameters have not changed, and the calculator type has not changed then the calculator is not modified.

This is resolved here: https://github.com/muon-spectroscopy-computational-project/pymuon-suite/commit/754456503b0fe4e047703545a3c997aef9058a91